pkg/apiserver/webhook/configmap_webhook.go

```diff
@@ -0,0 +1,94 @@
+	"net/http"
+	"regexp"
+	controllerclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// ConfigMapValidator is used to validate Config Map.
+type ConfigMapValidator struct {
+	Client  controllerclient.Client
+	Log     logr.Logger
+	decoder *admission.Decoder
+	Config  config.ControllerConfig
+	Npr     controllers.NetworkPolicyReconciler
+}
+
+// Handle handles admission request for config map
+func (v *ConfigMapValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	v.Log.V(1).Info("Received admission webhook request", "Name", req.Name, "Operation", req.Operation)
+	switch req.Operation {
+	case admissionv1.Create:
+		return v.validateCreate(req)
+	case admissionv1.Update:
+		return v.validateUpdate(req)
+	case admissionv1.Delete:
+		return v.validateDelete(req)
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("invalid admission webhook request"))
+	}
+}
+
+// InjectDecoder injects the decoder.
+func (v *ConfigMapValidator) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// validateCreate
+func (v *ConfigMapValidator) validateCreate(req admission.Request) admission.Response {
+	return admission.Allowed("")
+}
+
+// validateUpdate
+func (v *ConfigMapValidator) validateUpdate(req admission.Request) admission.Response {
+	newConfigMap := &corev1.ConfigMap{}
+	oldConfigMap := &corev1.ConfigMap{}
+	err := v.decoder.Decode(req, newConfigMap)
+	if err != nil {
+		v.Log.Error(err, "Failed to decode configMap", "ConfigMapValidator", req.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	v.Log.V(1).Info("new", "CloudResourcePrefix", newConfigMap.Data["CloudResourcePrefix"])
+	if req.OldObject.Raw != nil {
+		if err := json.Unmarshal(req.OldObject.Raw, &oldConfigMap); err != nil {
+			v.Log.Error(err, "Failed to decode old configMap", "ConfigMapValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+	v.Log.V(1).Info("old", "CloudResourcePrefix", oldConfigMap.Data["CloudResourcePrefix"])
+	if newConfigMap.Data["CloudResourcePrefix"] == oldConfigMap.Data["CloudResourcePrefix"] {
+		v.Log.V(1).Info("CloudResourcePrefix are equal", "newData", newConfigMap.Data["CloudResourcePrefix"], "oldDate", oldConfigMap.Data["CloudResourcePrefix"])
+		return admission.Allowed("")
+	}
+	if v.Npr.IsCloudResourceCreated() {
+		return admission.Denied(fmt.Sprintf("already cloud resource is created with CloudResourcePrefix %s. Please delete the resource and try or use the same CloudResourcePrefix", oldConfigMap.Data["CloudResourcePrefix"]))
+	}
+	if len(newConfigMap.Data["CloudResourcePrefix"]) > 0 {
+		reg := regexp.MustCompile(`^[a-zA-Z0-9]+([_ -]?[a-zA-Z0-9])*$`)
+		valid := reg.MatchString(newConfigMap.Data["CloudResourcePrefix"])
+		if !valid {
+			return admission.Denied(fmt.Sprintf("invalid CloudResourcePrefix %s", newConfigMap.Data["CloudResourcePrefix"]))
```

Comment:
The user should know why it is an error and what is the expected naming convention.

Reply:
We can provide it in the documentation because the naming rule is large

Reply:
invalid CloudResourcePrefix %s, only alphanumeric and '-' characters are allowed.

Reply:
Solved